### PR TITLE
Reflect function definitions in LFSC.

### PR DESCRIPTION
**Pull request by abdoo8080**
This PR modifies the LFSC proofs to print non-recursive function definitions as LFSC definitions instead of declarations and assumptions. The definitions are passed to the LFSC postprocessor and printer in a type-safe manner by adding a global scope in the proof node introducing the definitions. Since LFSC definitions are just macros, the LFSC postprocessor replaces applications of the ASSUME rule for definitions with REFL.

### CODE REVIEW

File: src/smt/proof_manager.cpp [hunk 1]
```diff
@@ -152,18 +151,56 @@ std::shared_ptr<ProofNode> PfManager::connectProofToAssertions(
...
+    {
+      Trace("smt-proof")
+          << "SolverEngine::connectProofToAssertions(): make split scope...\n";
+      // To support proof pruning for nested scopes, we need to:
+      // 1. Minimize assertions of closed unified scope.
+      std::vector<Node> unifiedAssertions;
+      getAssertions(as, unifiedAssertions);
+      Pf pf = d_pnm->mkScope(
+          pfn, unifiedAssertions, true, options().proof.proofPruneInput);
+      // 2. Extract minimum unified assertions from the scope node.
```
Comment (ajreynol) [inline src/smt/proof_manager.cpp:182]: Assert that pf has rule SCOPE here

File: src/smt/proof_manager.cpp [hunk 2]
```diff
@@ -305,5 +340,28 @@ void PfManager::getAssertions(Assertions& as,
   }
 }
 
+void PfManager::getDefinitionsAndAssertions(Assertions& as,
+                                            std::vector<Node>& definitions,
+                                            std::vector<Node>& assertions)
+{
+  for (const Node& d : as.getAssertionListDefinitions())
```
Comment (ajreynol) [inline src/smt/proof_manager.cpp:347]: Can you store the assertion list definitions into a local before you iterate? I'm not entirely sure how this compile (e.g. if the assertion list is retrieved again on every iteration).

File: src/smt/proof_manager.cpp [hunk 3]
```diff
@@ -305,5 +340,28 @@ void PfManager::getAssertions(Assertions& as,
...
+  {
+    // Keep treating (mutually) recursive functions as declarations +
+    // assertions.
+    if (d.getKind() == kind::EQUAL)
+    {
+      definitions.push_back(d);
+    }
+  }
+  for (const Node& a : as.getAssertionList())
```
Comment (ajreynol) [inline src/smt/proof_manager.cpp:356]: Ditto.

**Review by ajreynol — CHANGES_REQUESTED** [review]
This looks great, thanks for doing this.

Just a few minor requests.

File: src/smt/proof_manager.cpp (hunk 2 above)
Comment (abdoo8080) [reply src/smt/proof_manager.cpp]: hmm... Then I am kinda surprised you didn't complain about line 361 😅

**Review by ajreynol — APPROVED** [review]

**Pull request merged**